### ISSUE TITLE
Don't default to 3-CNOT decomposition when jitting for two-qubit unitary

### ISSUE THREAD

**Issue by lillian542** [PennyLaneAI/pennylane#3574]
### Feature details

Compatibility with jitting for decomposing two-qubit unitaries is added in #3569. In that PR, it defaults to the general solution (3 CNOT decomposition) when jitting. It would be better if it were possible to decompose into the optimal number of gates instead, to limit unnecessary gates in decompositions.

### Implementation

Either make `_compute_num_cnots` and the conditions in the other CNOT decomposition functions jit-compatible, or use `jax.pure_callback` to evaluate the number of CNOTS. Solution without `pure_callback` is preferred.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by astralcai** [comment]
With `capture` and `qjit`, the optimal decomposition is used. We only default to the 3 CNOT version for `jax.jit`. Is that sufficient?

**Comment by lillian542** [comment]
> With capture and qjit, the optimal decomposition is used. We only default to the 3 CNOT version for jax.jit. Is that sufficient?

Yes, that's plenty good enough 😅 What we tried to do two years ago wasn't feasible with the infrastructure we had at the time.
